### DISCUSSION TITLE
Content version history

### DISCUSSION THREAD

**Discussion by andrerav** [OrchardCMS/OrchardCore#11601] [Q&A]
Hi,
I can't seem to figure out how to view the version history of a content item and compare changes with previous versions. Where can I find this functionality?
Regards,
Andreas

**Comment by hishamco** [comment]
All the versions are stored in the database and managed by OC, I don't think there's a functionaly similar to what you asked for, nothing but you can have a look to AuditTrail module,I remember there's a view to display the diff, you can use the same pattern applied on versions

@domonkosgabor correct me if I'm wrong about diff view, I think you demo it If I'm not wrong

**Reply by andrerav** [reply]
The audit trail module seems to provide this functionality, but with a couple of big weaknesses that makes its usefulness very limited:
- It has to be explicitly enabled on a per-content-type basis
- There is no history available for a content item from any point in time _before_ the audit module was enabled (and auditing was enabled for the corresponding content type)
- I can't use it to restore previous versions created/modified before enabling the audit module

I am just looking for a way to view the internal version history of any content item, see who made the change, see the diff, and an option to restore a previous version. Seems like a pretty basic feature for a CMS?

**Reply by andrerav** [reply]
Here's a concrete example:

![image](https://user-images.githubusercontent.com/3635018/165546865-66b1a79f-89cb-46f7-bbb1-8a6857112d90.png)

Where is version 1, 2 and 3?

**Reply by hishamco** [reply]
Almost the CMSs I saw focusing on the latest version, but this a good feature for community feature, let us heard the folks feedback. In case you are interested to make this happen we can work on it in [Orchard Core Contrib](https://github.com/OrchardCoreContrib/OrchardCoreContrib.Modules)

**Reply by hishamco** [reply]
> Where is version 1, 2 and 3?

I will let @domonkosgabor reply to this coz he is the guys behind this if I'm not wrong, but again AuditTrail developed for a reason ans certain features, what you asking for is something custom

**Reply by domonkosgabor** [reply]
@andrerav 
"It has to be explicitly enabled on a per-content-type basis." We don't want to store Audit Trail records for every content type by default. If you know that you will need the version history of a given content type, you have to enable it. That doesn't make sense to store the history of all content items by default without the need for them.

"There is no history available for a content item from any point in time before the audit module was enabled (and auditing was enabled for the corresponding content type)" Yes if the Audit Trail module is not enabled and you haven't set a content type to track, you will not get a version history for any content item.

"I can't use it to restore previous versions created/modified before enabling the audit module" Same as above.

**Comment by Skrypt** [comment]
The version history can be seen from the Audit Trail module and there is a "diff" view that can be seen as @hishamco is saying. The version 1,2 and 3 are probably not there because the Audit Trail module was not enabled at that time.
